File: answers/leetcode/Simplify Path/Simplify Path.java
class Solution {
    public String simplifyPath(String path) {
        String[] folders = path.split("/");
        Stack<String> stack = new Stack();

        for (String folder : folders) {
            switch (folder) {
                case "":
                case ".":
                    break;
                case "..":
                    if (!stack.empty()) {
                        stack.pop();
                    }
                    break;
                default:
                    stack.push(folder);
                    break;
            }
        }

        return "/" + String.join("/", new LinkedList<>(stack));
    }
}
//Runtime: 6 ms, faster than 49.58% of Java online submissions for Simplify Path.
//Memory Usage: 38.9 MB, less than 94.06% of Java online submissions for Simplify Path.

File: answers/leetcode/Simplify Path/Simplify Path.py
class Solution(object):
    def simplifyPath(self, path):
        """
        :type path: str
        :rtype: str
        """

        path_list = []
        prev = 0
        for i in range(len(path) + 1):
            if len(path) == i or '/' == path[i]:
                if prev + 2 == i and '.' == path[i - 2] and '.' == path[i - 1]:
                    if len(path_list):
                        path_list.pop()
                elif i - prev > 0 and not (prev + 1 == i and '.' == path[prev]):
                    path_list.append(path[prev : (i if i < len(path) else i + 1)])
                prev = i + 1
        return '/' + '/'.join(path_list)
